ccl/multiregionccl: TestColdStartLatency failed

ccl/multiregionccl.TestColdStartLatency [failed](https://tanzanite.cluster.engflow.com/invocations/default/e71abb08-5df1-4731-8696-95790f0b9cf2?testReportRun=11&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [c9c3cc5f3c3a4a6ab556f4b9d5b6ec0381901bdb](https://github.com/cockroachdb/cockroach/commits/c9c3cc5f3c3a4a6ab556f4b9d5b6ec0381901bdb):


```
        	            	deadline below read timestamp is nonsensical; txn has would have no chance to commit. Deadline: 1708926205.478454608,0. Read timestamp: 1708926213.126141218,0 Previous Deadline: 0,0.
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).UpdateDeadline
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:862
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage.(*Storage).createInstanceRow.func1.1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage/instancestorage.go:263
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:1049
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.runTxn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1092
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1055
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1030
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage.(*Storage).createInstanceRow.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage/instancestorage.go:253
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage.(*Storage).createInstanceRow
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage/instancestorage.go:308
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage.(*Storage).CreateInstance
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage/instancestorage.go:183
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).preStart.func3
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1535
        	            	  | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetryEx[...]
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:142
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).preStart
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1518
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*SQLServerWrapper).PreStart
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/tenant.go:819
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*SQLServerWrapper).Start
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/tenant.go:1027
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*testServer).StartTenant
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/testserver.go:1756
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl.TestColdStartLatency.func8
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl/cold_start_latency_test.go:325
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1650
        	            	Wraps: (3) deadline below read timestamp is nonsensical; txn has would have no chance to commit. Deadline: 1708926205.478454608,0. Read timestamp: 1708926213.126141218,0 Previous Deadline: 0,0.
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestColdStartLatency
    cold_start_latency_test.go:353: done 4 region=us-east1,az=b 1.804317934s
    cold_start_latency_test.go:353: done 5 region=us-east1,az=c 2.396299957s
    cold_start_latency_test.go:353: done 6 region=us-west1,az=a 4.076042532s
    cold_start_latency_test.go:353: done 7 region=us-west1,az=b 3.871147733s
    cold_start_latency_test.go:353: done 8 region=us-west1,az=c 4.139402078s
    cold_start_latency_test.go:376: result [region=europe-west1,az=a region=europe-west1,az=b region=europe-west1,az=c region=us-east1,az=a region=us-east1,az=b region=us-east1,az=c region=us-west1,az=a region=us-west1,az=b region=us-west1,az=c] [1m1.703536868s 641.584692ms 660.292768ms 0s 1.804317934s 2.396299957s 4.076042532s 3.871147733s 4.139402078s]
    cold_start_latency_test.go:378: -- test log scope end --
test logs left over in: outputs.zip/logTestColdStartLatency3065315420
--- FAIL: TestColdStartLatency (203.53s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=11</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestColdStartLatency.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>



Jira issue: CRDB-36221

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -514,8 +514,7 @@ func (r *refreshInstanceSessionListener) OnSessionDeleted(
 			}
 			if _, err := r.cfg.sqlInstanceStorage.CreateNodeInstance(
 				ctx,
-				s.ID(),
-				s.Expiration(),
+				s,
 				r.cfg.AdvertiseAddr,
 				r.cfg.SQLAdvertiseAddr,
 				r.cfg.Locality,
@@ -1516,8 +1515,7 @@ func (s *SQLServer) preStart(
 				// Write/acquire our instance row.
 				return s.sqlInstanceStorage.CreateNodeInstance(
 					ctx,
-					session.ID(),
-					session.Expiration(),
+					session,
 					s.cfg.AdvertiseAddr,
 					s.cfg.SQLAdvertiseAddr,
 					s.distSQLServer.Locality,
@@ -1527,8 +1525,7 @@ func (s *SQLServer) preStart(
 			}
 			return s.sqlInstanceStorage.CreateInstance(
 				ctx,
-				session.ID(),
-				session.Expiration(),
+				session,
 				s.cfg.AdvertiseAddr,
 				s.cfg.SQLAdvertiseAddr,
 				s.distSQLServer.Locality,

--- a/pkg/sql/sqlinstance/instancestorage/BUILD.bazel
+++ b/pkg/sql/sqlinstance/instancestorage/BUILD.bazel
@@ -80,6 +80,7 @@ go_test(
         "//pkg/sql/sqlinstance",
         "//pkg/sql/sqlliveness",
         "//pkg/sql/sqlliveness/slstorage",
+        "//pkg/sql/sqlliveness/sqllivenesstestutils",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
         "//pkg/testutils/sqlutils",

--- a/pkg/sql/sqlinstance/instancestorage/instancereader_test.go
+++ b/pkg/sql/sqlinstance/instancestorage/instancereader_test.go
@@ -26,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/sqllivenesstestutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
@@ -72,7 +73,7 @@ func TestReader(t *testing.T) {
 	})
 	t.Run("read-without-waiting", func(t *testing.T) {
 		storage, slStorage, clock, reader := setup(t)
-		sessionID := makeSession()
+		session := makeSession()
 		const rpcAddr = "rpcAddr"
 		const sqlAddr = "sqlAddr"
 		binaryVersion := roachpb.Version{Major: 23, Minor: 2}
@@ -80,10 +81,11 @@ func TestReader(t *testing.T) {
 		// Set a high enough expiration to ensure the session stays
 		// live through the test.
 		const expiration = 10 * time.Minute
-		sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
-		instance, err := storage.CreateInstance(ctx, sessionID, sessionExpiry, rpcAddr, sqlAddr, locality, binaryVersion)
+		session.StartTS = clock.Now()
+		session.ExpTS = session.StartTS.Add(expiration.Nanoseconds(), 0)
+		instance, err := storage.CreateInstance(ctx, session, rpcAddr, sqlAddr, locality, binaryVersion)
 		require.NoError(t, err)
-		err = slStorage.Insert(ctx, sessionID, sessionExpiry)
+		err = slStorage.Insert(ctx, session.ID(), session.Expiration())
 		require.NoError(t, err)
 		reader.Start(ctx, instance)
 
@@ -99,7 +101,7 @@ func TestReader(t *testing.T) {
 		storage, slStorage, clock, reader := setup(t)
 		reader.Start(ctx, sqlinstance.InstanceInfo{})
 		require.NoError(t, reader.WaitForStarted(ctx))
-		sessionID := makeSession()
+		session := makeSession()
 		const rpcAddr = "rpcAddr"
 		const sqlAddr = "sqlAddr"
 		binaryVersion := roachpb.Version{Major: 25, Minor: 3}
@@ -108,12 +110,13 @@ func TestReader(t *testing.T) {
 		// live through the test.
 		const expiration = 10 * time.Minute
 		{
-			sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
-			instance, err := storage.CreateInstance(ctx, sessionID, sessionExpiry, rpcAddr, sqlAddr, locality, binaryVersion)
+			session.StartTS = clock.Now()
+			session.ExpTS = session.StartTS.Add(expiration.Nanoseconds(), 0)
+			instance, err := storage.CreateInstance(ctx, session, rpcAddr, sqlAddr, locality, binaryVersion)
 			if err != nil {
 				t.Fatal(err)
 			}
-			err = slStorage.Insert(ctx, sessionID, sessionExpiry)
+			err = slStorage.Insert(ctx, session.ID(), session.Expiration())
 			if err != nil {
 				t.Fatal(err)
 			}
@@ -150,7 +153,7 @@ func TestReader(t *testing.T) {
 		instanceIDs := []base.SQLInstanceID{1, 2, 3}
 		rpcAddresses := []string{"addr1", "addr2", "addr3"}
 		sqlAddresses := []string{"addr4", "addr5", "addr6"}
-		sessionIDs := []sqlliveness.SessionID{makeSession(), makeSession(), makeSession()}
+		sessions := []*sqllivenesstestutils.FakeSession{makeSession(), makeSession(), makeSession()}
 		localities := []roachpb.Locality{
 			{Tiers: []roachpb.Tier{{Key: "region", Value: "region1"}}},
 			{Tiers: []roachpb.Tier{{Key: "region", Value: "region2"}}},
@@ -229,11 +232,15 @@ func TestReader(t *testing.T) {
 		}
 
 		expectationsFromOffset := func(offset int) expectations {
+			sessionIDs := make([]sqlliveness.SessionID, 0, len(sessions[offset:]))
+			for _, session := range sessions[offset:] {
+				sessionIDs = append(sessionIDs, session.ID())
+			}
 			return expectations{
 				instanceIDs:    instanceIDs[offset:],
 				rpcAddresses:   rpcAddresses[offset:],
 				sqlAddresses:   sqlAddresses[offset:],
-				sessionIDs:     sessionIDs[offset:],
+				sessionIDs:     sessionIDs,
 				localities:     localities[offset:],
 				binaryVersions: binaryVersions[offset:],
 			}
@@ -242,12 +249,13 @@ func TestReader(t *testing.T) {
 		{
 			// Set up mock data within instance and session storage.
 			for index, rpcAddr := range rpcAddresses {
-				sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
-				_, err := storage.CreateInstance(ctx, sessionIDs[index], sessionExpiry, rpcAddr, sqlAddresses[index], localities[index], binaryVersions[index])
+				sessions[index].StartTS = clock.Now()
+				sessions[index].ExpTS = sessions[index].StartTS.Add(expiration.Nanoseconds(), 0)
+				_, err := storage.CreateInstance(ctx, sessions[index], rpcAddr, sqlAddresses[index], localities[index], binaryVersions[index])
 				if err != nil {
 					t.Fatal(err)
 				}
-				err = slStorage.Insert(ctx, sessionIDs[index], sessionExpiry)
+				err = slStorage.Insert(ctx, sessions[index].SessionID, sessions[index].Expiration())
 				if err != nil {
 					t.Fatal(err)
 				}
@@ -260,7 +268,7 @@ func TestReader(t *testing.T) {
 
 		// Release an instance and verify only active instances are returned.
 		{
-			err := storage.ReleaseInstance(ctx, sessionIDs[0], instanceIDs[0])
+			err := storage.ReleaseInstance(ctx, sessions[0].ID(), instanceIDs[0])
 			if err != nil {
 				t.Fatal(err)
 			}
@@ -271,7 +279,7 @@ func TestReader(t *testing.T) {
 
 		// Verify instances with expired sessions are filtered out.
 		{
-			err := slStorage.Delete(ctx, sessionIDs[1])
+			err := slStorage.Delete(ctx, sessions[1].ID())
 			if err != nil {
 				t.Fatal(err)
 			}
@@ -284,14 +292,15 @@ func TestReader(t *testing.T) {
 		// the latest instance information is returned. This heuristic is used
 		// when instance information isn't released correctly prior to SQL instance shutdown.
 		{
-			sessionID := makeSession()
+			session := makeSession()
 			locality := roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: "region4"}}}
-			sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
-			instance, err := storage.CreateInstance(ctx, sessionID, sessionExpiry, rpcAddresses[2], sqlAddresses[2], locality, binaryVersions[2])
+			session.StartTS = clock.Now()
+			session.ExpTS = session.StartTS.Add(expiration.Nanoseconds(), 0)
+			instance, err := storage.CreateInstance(ctx, session, rpcAddresses[2], sqlAddresses[2], locality, binaryVersions[2])
 			if err != nil {
 				t.Fatal(err)
 			}
-			err = slStorage.Insert(ctx, sessionID, sessionExpiry)
+			err = slStorage.Insert(ctx, session.ID(), session.Expiration())
 			if err != nil {
 				t.Fatal(err)
 			}
@@ -300,7 +309,7 @@ func TestReader(t *testing.T) {
 					[]base.SQLInstanceID{instance.InstanceID}, /* instanceIDs */
 					[]string{rpcAddresses[2]},                 /* rpcAddresses */
 					[]string{sqlAddresses[2]},                 /* sqlAddresses */
-					[]sqlliveness.SessionID{sessionID},        /* sessionIDs */
+					[]sqlliveness.SessionID{session.ID()},     /* sessions */
 					[]roachpb.Locality{locality},              /* localities */
 					[]roachpb.Version{binaryVersions[2]},      /* binaryVersions */
 				})
@@ -318,7 +327,7 @@ func TestReader(t *testing.T) {
 		instanceIDs := [...]base.SQLInstanceID{1, 2, 3}
 		rpcAddresses := [...]string{"addr1", "addr2", "addr3"}
 		sqlAddresses := [...]string{"addr4", "addr5", "addr6"}
-		sessionIDs := [...]sqlliveness.SessionID{makeSession(), makeSession(), makeSession()}
+		sessions := [...]*sqllivenesstestutils.FakeSession{makeSession(), makeSession(), makeSession()}
 		localities := [...]roachpb.Locality{
 			{Tiers: []roachpb.Tier{{Key: "region", Value: "region1"}}},
 			{Tiers: []roachpb.Tier{{Key: "region", Value: "region2"}}},
@@ -330,12 +339,13 @@ func TestReader(t *testing.T) {
 		{
 			// Set up mock data within instance and session storage.
 			for index, rpcAddr := range rpcAddresses {
-				sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
-				_, err := storage.CreateInstance(ctx, sessionIDs[index], sessionExpiry, rpcAddr, sqlAddresses[index], localities[index], binaryVersions[index])
+				sessions[index].StartTS = clock.Now()
+				sessions[index].ExpTS = sessions[index].StartTS.Add(expiration.Nanoseconds(), 0)
+				_, err := storage.CreateInstance(ctx, sessions[index], rpcAddr, sqlAddresses[index], localities[index], binaryVersions[index])
 				if err != nil {
 					t.Fatal(err)
 				}
-				err = slStorage.Insert(ctx, sessionIDs[index], sessionExpiry)
+				err = slStorage.Insert(ctx, sessions[index].ID(), sessions[index].Expiration())
 				if err != nil {
 					t.Fatal(err)
 				}
@@ -367,7 +377,7 @@ func TestReader(t *testing.T) {
 
 		// Verify request for released instance data results in an error.
 		{
-			err := storage.ReleaseInstance(ctx, sessionIDs[0], instanceIDs[0])
+			err := storage.ReleaseInstance(ctx, sessions[0].ID(), instanceIDs[0])
 			if err != nil {
 				t.Fatal(err)
 			}
@@ -381,7 +391,7 @@ func TestReader(t *testing.T) {
 		}
 		// Verify request for instance with expired session results in an error.
 		{
-			err := slStorage.Delete(ctx, sessionIDs[1])
+			err := slStorage.Delete(ctx, sessions[1].ID())
 			if err != nil {
 				t.Fatal(err)
 			}

--- a/pkg/sql/sqlinstance/instancestorage/instancestorage.go
+++ b/pkg/sql/sqlinstance/instancestorage/instancestorage.go
@@ -156,15 +156,14 @@ func NewStorage(
 // associates it with its SQL address and session information.
 func (s *Storage) CreateNodeInstance(
 	ctx context.Context,
-	sessionID sqlliveness.SessionID,
-	sessionExpiration hlc.Timestamp,
+	session sqlliveness.Session,
 	rpcAddr string,
 	sqlAddr string,
 	locality roachpb.Locality,
 	binaryVersion roachpb.Version,
 	nodeID roachpb.NodeID,
 ) (instance sqlinstance.InstanceInfo, _ error) {
-	return s.createInstanceRow(ctx, sessionID, sessionExpiration, rpcAddr, sqlAddr, locality, binaryVersion, nodeID)
+	return s.createInstanceRow(ctx, session, rpcAddr, sqlAddr, locality, binaryVersion, nodeID)
 }
 
 const noNodeID = 0
@@ -173,14 +172,13 @@ const noNodeID = 0
 // associates it with its SQL address and session information.
 func (s *Storage) CreateInstance(
 	ctx context.Context,
-	sessionID sqlliveness.SessionID,
-	sessionExpiration hlc.Timestamp,
+	session sqlliveness.Session,
 	rpcAddr string,
 	sqlAddr string,
 	locality roachpb.Locality,
 	binaryVersion roachpb.Version,
 ) (instance sqlinstance.InstanceInfo, _ error) {
-	return s.createInstanceRow(ctx, sessionID, sessionExpiration, rpcAddr, sqlAddr, locality, binaryVersion, noNodeID)
+	return s.createInstanceRow(ctx, session, rpcAddr, sqlAddr, locality, binaryVersion, noNodeID)
 }
 
 // ReleaseInstance deallocates the instance id iff it is currently owned by the
@@ -225,8 +223,7 @@ func (s *Storage) ReleaseInstance(
 
 func (s *Storage) createInstanceRow(
 	ctx context.Context,
-	sessionID sqlliveness.SessionID,
-	sessionExpiration hlc.Timestamp,
+	session sqlliveness.Session,
 	rpcAddr string,
 	sqlAddr string,
 	locality roachpb.Locality,
@@ -236,11 +233,11 @@ func (s *Storage) createInstanceRow(
 	if len(sqlAddr) == 0 || len(rpcAddr) == 0 {
 		return sqlinstance.InstanceInfo{}, errors.AssertionFailedf("missing sql or rpc address information for instance")
 	}
-	if len(sessionID) == 0 {
+	if len(session.ID()) == 0 {
 		return sqlinstance.InstanceInfo{}, errors.AssertionFailedf("no session information for instance")
 	}
 
-	region, _, err := slstorage.UnsafeDecodeSessionID(sessionID)
+	region, _, err := slstorage.UnsafeDecodeSessionID(session.ID())
 	if err != nil {
 		return sqlinstance.InstanceInfo{}, errors.Wrap(err, "unable to determine region for sql_instance")
 	}
@@ -260,7 +257,7 @@ func (s *Storage) createInstanceRow(
 
 			// Set the transaction deadline to the session expiration to ensure
 			// transaction commits before the session expires.
-			err = txn.UpdateDeadline(ctx, sessionExpiration)
+			err = txn.UpdateDeadline(ctx, session.Expiration())
 			if err != nil {
 				return err
 			}
@@ -284,7 +281,7 @@ func (s *Storage) createInstanceRow(
 
 			b := txn.NewBatch()
 
-			value, err := s.rowCodec.encodeValue(rpcAddr, sqlAddr, sessionID, locality, binaryVersion)
+			value, err := s.rowCodec.encodeValue(rpcAddr, sqlAddr, session.ID(), locality, binaryVersion)
 			if err != nil {
 				return err
 			}
@@ -313,7 +310,7 @@ func (s *Storage) createInstanceRow(
 				InstanceID:      instanceID,
 				InstanceRPCAddr: rpcAddr,
 				InstanceSQLAddr: sqlAddr,
-				SessionID:       sessionID,
+				SessionID:       session.ID(),
 				Locality:        locality,
 				BinaryVersion:   binaryVersion,
 			}, err
@@ -333,7 +330,7 @@ func (s *Storage) createInstanceRow(
 		// every region, then writing to the local region. Allocating globally
 		// would require one round trip for reading and one round trip for
 		// writes.
-		if err := s.generateAvailableInstanceRows(ctx, [][]byte{region}, sessionExpiration); err != nil {
+		if err := s.generateAvailableInstanceRows(ctx, [][]byte{region}, session.Expiration()); err != nil {
 			log.Warningf(ctx, "failed to generate available instance rows: %v", err)
 		}
 	}
@@ -603,7 +600,7 @@ func (s *Storage) readRegionsFromSystemDatabase(
 // them in the sql_instances table. When instance IDs are pre-allocated, all
 // other fields in that row will be NULL.
 func (s *Storage) generateAvailableInstanceRowsWithTxn(
-	ctx context.Context, regions [][]byte, sessionExpiration hlc.Timestamp, txn *kv.Txn, commit bool,
+	ctx context.Context, regions [][]byte, txn *kv.Txn, commit bool,
 ) error {
 	target := int(PreallocatedCount.Get(&s.settings.SV))
 	// Figure out which regions are down, so that we can skip them in the allocation
@@ -653,7 +650,7 @@ func (s *Storage) generateAvailableInstanceRows(
 ) error {
 	ctx = multitenant.WithTenantCostControlExemption(ctx)
 	return s.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
-		return s.generateAvailableInstanceRowsWithTxn(ctx, regions, sessionExpiration, txn, true)
+		return s.generateAvailableInstanceRowsWithTxn(ctx, regions, txn, true)
 	})
 }
 

--- a/pkg/sql/sqlinstance/instancestorage/instancestorage_test.go
+++ b/pkg/sql/sqlinstance/instancestorage/instancestorage_test.go
@@ -31,6 +31,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/sqllivenesstestutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
@@ -46,12 +47,12 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func makeSession() sqlliveness.SessionID {
-	session, err := slstorage.MakeSessionID(enum.One, uuid.MakeV4())
+func makeSession() *sqllivenesstestutils.FakeSession {
+	sessionID, err := slstorage.MakeSessionID(enum.One, uuid.MakeV4())
 	if err != nil {
 		panic(err)
 	}
-	return session
+	return &sqllivenesstestutils.FakeSession{SessionID: sessionID}
 }
 
 // TestStorage verifies that instancestorage stores and retrieves SQL instance data correctly.
@@ -89,14 +90,16 @@ func TestStorage(t *testing.T) {
 		stopper, storage, _, clock := setup(t)
 		defer stopper.Stop(ctx)
 		const id = base.SQLInstanceID(1)
-		sessionID := makeSession()
+		session := makeSession()
 		const rpcAddr = "rpcAddr"
 		const sqlAddr = "sqlAddr"
 		locality := roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: "test"}, {Key: "az", Value: "a"}}}
 		binaryVersion := roachpb.Version{Major: 28, Minor: 4}
 		const expiration = time.Minute
 		{
-			instance, err := storage.CreateInstance(ctx, sessionID, clock.Now().Add(expiration.Nanoseconds(), 0), rpcAddr, sqlAddr, locality, binaryVersion)
+			session.StartTS = clock.Now()
+			session.ExpTS = session.StartTS.Add(expiration.Nanoseconds(), 0)
+			instance, err := storage.CreateInstance(ctx, session, rpcAddr, sqlAddr, locality, binaryVersion)
 			require.NoError(t, err)
 			require.Equal(t, id, instance.InstanceID)
 		}
@@ -107,6 +110,7 @@ func TestStorage(t *testing.T) {
 		stopper, storage, slStorage, clock := setup(t)
 		defer stopper.Stop(ctx)
 
+		sessionStart := clock.Now()
 		sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
 
 		makeInstance := func(id int) sqlinstance.InstanceInfo {
@@ -115,7 +119,7 @@ func TestStorage(t *testing.T) {
 				InstanceID:      base.SQLInstanceID(id),
 				InstanceSQLAddr: fmt.Sprintf("sql-addr-%d", id),
 				InstanceRPCAddr: fmt.Sprintf("rpc-addr-%d", id),
-				SessionID:       makeSession(),
+				SessionID:       makeSession().ID(),
 				Locality:        roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: fmt.Sprintf("region-%d", id)}}},
 				BinaryVersion:   roachpb.Version{Major: 22, Minor: int32(id)},
 			}
@@ -130,7 +134,10 @@ func TestStorage(t *testing.T) {
 				require.NoError(t, slStorage.Insert(ctx, instance.SessionID, sessionExpiry))
 			}
 
-			created, err := storage.CreateInstance(ctx, instance.SessionID, sessionExpiry, instance.InstanceRPCAddr, instance.InstanceSQLAddr, instance.Locality, instance.BinaryVersion)
+			session := &sqllivenesstestutils.FakeSession{SessionID: instance.SessionID,
+				StartTS: sessionStart,
+				ExpTS:   sessionExpiry}
+			created, err := storage.CreateInstance(ctx, session, instance.InstanceRPCAddr, instance.InstanceSQLAddr, instance.Locality, instance.BinaryVersion)
 			require.NoError(t, err)
 
 			require.Equal(t, instance, created)
@@ -268,10 +275,12 @@ func TestSQLAccess(t *testing.T) {
 	var locality roachpb.Locality
 	var binaryVersion roachpb.Version
 	require.NoError(t, locality.Set(tierStr))
+	session := makeSession()
+	session.StartTS = clock.Now()
+	session.ExpTS = session.StartTS.Add(expiration.Nanoseconds(), 0)
 	instance, err := storage.CreateInstance(
 		ctx,
-		makeSession(),
-		clock.Now().Add(expiration.Nanoseconds(), 0),
+		session,
 		"rpcAddr",
 		"sqlAddr",
 		locality,
@@ -402,11 +411,12 @@ func TestConcurrentCreateAndRelease(t *testing.T) {
 		sqlAddr         = "sqlAddr"
 		expiration      = time.Minute
 	)
-	sessionID := makeSession()
+	session := makeSession()
+	session.StartTS = clock.Now()
+	session.ExpTS = session.StartTS.Add(expiration.Nanoseconds(), 0)
 	locality := roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: "test-region"}}}
 	binaryVersion := roachpb.Version{Major: 23, Minor: 4}
-	sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
-	err := slStorage.Insert(ctx, sessionID, sessionExpiry)
+	err := slStorage.Insert(ctx, session.ID(), session.Expiration())
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -425,12 +435,12 @@ func TestConcurrentCreateAndRelease(t *testing.T) {
 			t.Helper()
 			state.Lock()
 			defer state.Unlock()
-			sessionExpiry = clock.Now().Add(expiration.Nanoseconds(), 0)
-			_, _, err = slStorage.Update(ctx, sessionID, sessionExpiry)
+			session.ExpTS = clock.Now().Add(expiration.Nanoseconds(), 0)
+			_, _, err = slStorage.Update(ctx, session.ID(), session.Expiration())
 			if err != nil {
 				t.Fatal(err)
 			}
-			instance, err := storage.CreateInstance(ctx, sessionID, sessionExpiry, rpcAddr, sqlAddr, locality, binaryVersion)
+			instance, err := storage.CreateInstance(ctx, session, rpcAddr, sqlAddr, locality, binaryVersion)
 			require.NoError(t, err)
 			if len(state.freeInstances) > 0 {
 				_, free := state.freeInstances[instance.InstanceID]
@@ -454,7 +464,7 @@ func TestConcurrentCreateAndRelease(t *testing.T) {
 			if i == -1 {
 				return
 			}
-			require.NoError(t, storage.ReleaseInstance(ctx, sessionID, i))
+			require.NoError(t, storage.ReleaseInstance(ctx, session.ID(), i))
 			state.freeInstances[i] = struct{}{}
 			delete(state.liveInstances, i)
 		}
@@ -493,7 +503,7 @@ func TestConcurrentCreateAndRelease(t *testing.T) {
 			} else {
 				require.Equal(t, rpcAddr, instanceInfo.InstanceRPCAddr)
 				require.Equal(t, sqlAddr, instanceInfo.InstanceSQLAddr)
-				require.Equal(t, sessionID, instanceInfo.SessionID)
+				require.Equal(t, session.ID(), instanceInfo.SessionID)
 				require.Equal(t, locality, instanceInfo.Locality)
 				require.Equal(t, binaryVersion, instanceInfo.BinaryVersion)
 				_, live := state.liveInstances[i]
@@ -557,7 +567,8 @@ func TestReclaimLoop(t *testing.T) {
 
 	// Expiration < ReclaimLoopInterval.
 	const expiration = 5 * time.Hour
-	sessionExpiry := clock.Now().Add(expiration.Nanoseconds(), 0)
+	sessionStart := clock.Now()
+	sessionExpiry := sessionStart.Add(expiration.Nanoseconds(), 0)
 
 	db := s.InternalDB().(descs.DB)
 	err := storage.RunInstanceIDReclaimLoop(ctx, s.AppStopper(), ts, db, func() hlc.Timestamp {
@@ -609,7 +620,11 @@ func TestReclaimLoop(t *testing.T) {
 	instanceIDs := [...]base.SQLInstanceID{1, 2}
 	rpcAddresses := [...]string{"addr1", "addr2"}
 	sqlAddresses := [...]string{"addr3", "addr4"}
-	sessionIDs := [...]sqlliveness.SessionID{makeSession(), makeSession()}
+	sessionIDs := [...]*sqllivenesstestutils.FakeSession{makeSession(), makeSession()}
+	for _, id := range sessionIDs {
+		id.StartTS = sessionStart
+		id.ExpTS = sessionExpiry
+	}
 	localities := [...]roachpb.Locality{
 		{Tiers: []roachpb.Tier{{Key: "region", Value: "region1"}}},
 		{Tiers: []roachpb.Tier{{Key: "region", Value: "region2"}}},
@@ -619,14 +634,14 @@ func TestReclaimLoop(t *testing.T) {
 	}
 
 	for i, id := range instanceIDs {
-		require.NoError(t, slStorage.Insert(ctx, sessionIDs[i], sessionExpiry))
+		require.NoError(t, slStorage.Insert(ctx, sessionIDs[i].ID(), sessionExpiry))
 		require.NoError(t, storage.CreateInstanceDataForTest(
 			ctx,
 			region,
 			id,
 			rpcAddresses[i],
 			sqlAddresses[i],
-			sessionIDs[i],
+			sessionIDs[i].ID(),
 			sessionExpiry,
 			localities[i],
 			binaryVersions[i],
@@ -662,7 +677,7 @@ func TestReclaimLoop(t *testing.T) {
 		case 0, 1:
 			require.Equal(t, rpcAddresses[i], instance.InstanceRPCAddr)
 			require.Equal(t, sqlAddresses[i], instance.InstanceSQLAddr)
-			require.Equal(t, sessionIDs[i], instance.SessionID)
+			require.Equal(t, sessionIDs[i].ID(), instance.SessionID)
 			require.Equal(t, localities[i], instance.Locality)
 			require.Equal(t, binaryVersions[i], instance.BinaryVersion)
 		default:

